pipelines/resource_pipeline.go

```diff
@@ -335,5 +336,41 @@ func ResourcePipeline() common.Resource {
 		Timeouts: &schema.ResourceTimeout{
 			Default: schema.DefaultTimeout(DefaultTimeout),
 		},
+		CustomizeDiff: func(ctx context.Context, d *schema.ResourceDiff) error {
+			// Allow changing catalog value in existing pipelines, but force recreation
+			// when switching between storage and catalog (or vice versa).
+			// This should only run on update, thus we skip this check if the ID is not known.
+			if d.Id() != "" {
+				storageChanged := d.HasChange("storage")
+				catalogChanged := d.HasChange("catalog")
+
+				// If both changed, it means we're switching between storage and catalog modes
+				if storageChanged && catalogChanged {
+					oldStorage, newStorage := d.GetChange("storage")
+					oldCatalog, newCatalog := d.GetChange("catalog")
+
+					// Force new if switching from storage to catalog
+					if oldStorage != "" && oldStorage != nil && newCatalog != "" && newCatalog != nil {
+						if err := d.ForceNew("catalog"); err != nil {
+							return err
+						}
+						if err := d.ForceNew("storage"); err != nil {
+							return err
+						}
+					}
+
+					// Force new if switching from catalog to storage
+					if oldCatalog != "" && oldCatalog != nil && newStorage != "" && newStorage != nil {
+						if err := d.ForceNew("catalog"); err != nil {
+							return err
+						}
+						if err := d.ForceNew("storage"); err != nil {
+							return err
+						}
+					}
```

Comment:
Shouldn't a change in either always result in recreation, unconditionally?

If not, it would be nice to include a comment stating why the logic is necessary.